src/Chapter4.hs

```diff
@@ -470,12 +484,20 @@ Applicatives can be found in many applications:
 
 Implement the Applicative instance for our 'Secret' data type from before.
 -}
+
+-- data Secret e a
+--     = Trap e
+--     | Reward a
+--     deriving (Show, Eq)
+
 instance Applicative (Secret e) where
-    pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+  pure :: a -> Secret e a
+  pure = Reward
 
-    (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+  (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
+  (Reward f) <*> a        = fmap f a
+  (Trap   e) <*> _        = Trap e
+  _          <*> (Trap e) = Trap e
```

Comment:
This case is actually redundant! See, if the left argument is `Trap`, this is covered by the previous line. And if it's `Reward`, the `Functor` instance for `Secret` already handles it, and it will return `Trap e` :slightly_smiling_face:

Reply:
💡 That's really nice! 😃